### PR TITLE
Create usergroups and add users to them automatically

### PR DESCRIPTION
**Pull request by dwillie**

### CODE REVIEW

File: pkgs/mkService.nix [hunk 1]
```diff
@@ -77,16 +77,25 @@ let
...
+
   execStartPre = optionalScript "${name}-prestart" (lib.concatStrings [
     (lib.optionalString user.create ''
       # Setup user
       if ! ${stdenv.glibc.bin}/bin/getent passwd ${user.name} > /dev/null; then
         ${remoteShadow}/bin/useradd --system --user-group ${user.name} --home ${if user.home == null then "/var/empty" else user.home}
       fi
+
+      # Ensure user groups initialised (if any) and then add user to groups (if any)
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/groupadd -f ${group}\n") user.groups))}
```
Comment (rimmington) [inline pkgs/mkService.nix]: I'm not sure it's a good idea to create groups by default; the user might intend to run the service under an existing system group. Instead, add a `userGroups` attr to `user` and create only those.

File: pkgs/mkService.nix [hunk 2]
```diff
@@ -77,16 +77,25 @@ let
...
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/usermod -a -G ${group} ${user.name}\n") user.groups))}
+
+      # Create user home
```
Comment (rimmington) [inline pkgs/mkService.nix]: I think `# Setup user` covers these, and in a lot of cases these comments will be followed by empty space.

File: pkgs/mkService.nix [hunk 3]
```diff
@@ -77,16 +77,25 @@ let
...
   execStartPre = optionalScript "${name}-prestart" (lib.concatStrings [
     (lib.optionalString user.create ''
       # Setup user
       if ! ${stdenv.glibc.bin}/bin/getent passwd ${user.name} > /dev/null; then
         ${remoteShadow}/bin/useradd --system --user-group ${user.name} --home ${if user.home == null then "/var/empty" else user.home}
       fi
+
+      # Ensure user groups initialised (if any) and then add user to groups (if any)
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/groupadd -f ${group}\n") user.groups))}
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/usermod -a -G ${group} ${user.name}\n") user.groups))}
```
Comment (rimmington) [inline pkgs/mkService.nix]: This isn't necessary; `chpst` (used to switch user) doesn't look at the groups a user is in.

File: pkgs/mkService.nix [hunk 4]
```diff
@@ -77,16 +77,25 @@ let
           RuntimeDirectory = systemdRequiredPaths runtimeDirs;
           RuntimeDirectoryMode = runtimeDirsMode;
        };
+
```
Comment (rimmington) [inline pkgs/mkService.nix]: :frowning_face:

**Review by rimmington — CHANGES_REQUESTED** [review]

File: pkgs/mkService.nix (hunk 1 above)
Comment (dwillie) [reply pkgs/mkService.nix]: So groups for system groups and userGroups for the groups we're creating ourselves?

File: pkgs/mkService.nix (hunk 3 above)
Comment (dwillie) [reply pkgs/mkService.nix]: If the user isn't a member of the group, how do we use it for restricting filesystem access?

File: pkgs/mkService.nix (hunk 1 above)
Comment (rimmington) [reply pkgs/mkService.nix]: Yes.

File: pkgs/mkService.nix (hunk 3 above)
Comment (rimmington) [reply pkgs/mkService.nix]: The group will be one of the executable's [supplementary group ids](http://unix.stackexchange.com/a/18290).

**Comment by dwillie** [comment]
@rimmington I think that addresses what you want, is it sane? Seems to work. Also, did you want me to use commit tags?

File: pkgs/mkService.nix [hunk 5]
```diff
@@ -100,6 +101,7 @@ let
     '') runtimeDirs)
     (lib.optionalString (preStartRootScript != "") "\n# Service prestart\n${preStartRootScript}")
   ]);
+
```
Comment (rimmington) [inline pkgs/mkService.nix]: :frowning_face:

File: pkgs/mkService.nix [hunk 6]
```diff
@@ -83,6 +83,7 @@ let
       if ! ${stdenv.glibc.bin}/bin/getent passwd ${user.name} > /dev/null; then
         ${remoteShadow}/bin/useradd --system --user-group ${user.name} --home ${if user.home == null then "/var/empty" else user.home}
       fi
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/groupadd -f ${group}\n") user.userGroups))}
```
Comment (rimmington) [inline pkgs/mkService.nix]: `lib.concatStrings (map f l)` ≡ `lib.concatMapStrings f l`

File: pkgs/mkService.nix [hunk 7]
```diff
@@ -113,7 +115,9 @@ let
     if user.name == "root"
       then exec
       else
-        let suf = lib.optionalString (user.groups != []) ":${lib.concatStringsSep ":" user.groups}";
+        let
+          allGroups = (builtins.concatLists [ user.groups user.userGroups ]);
```
Comment (rimmington) [inline pkgs/mkService.nix]: `builtins.concatLists [ as bs ]` ≡ `as ++ bs`

File: pkgs/mkService.nix [hunk 8]
```diff
@@ -113,7 +115,9 @@ let
...
+          suf = lib.optionalString (allGroups != []) ":${lib.concatStringsSep ":" allGroups}";
```
Comment (rimmington) [inline pkgs/mkService.nix]: `lib.optionalString (l != []) ":${lib.concatStringsSep ":" l}"` ≡ `lib.concatMapStrings (i: ":${i}") l`

(This one is my fault.)

**Review by rimmington — CHANGES_REQUESTED** [review]
I think this is the correct approach. I guess it'd be good to use commit tags, but I'm not really fussed.

Please also update `doc/mkService.md` with the new `userGroups` arg.

File: pkgs/mkService.nix [hunk 9]
```diff
@@ -83,6 +83,7 @@ let
       if ! ${stdenv.glibc.bin}/bin/getent passwd ${user.name} > /dev/null; then
         ${remoteShadow}/bin/useradd --system --user-group ${user.name} --home ${if user.home == null then "/var/empty" else user.home}
       fi
+      ${(lib.concatMapStrings (group: "${remoteShadow}/bin/groupadd -f ${group}\n") user.userGroups)}
```
Comment (rimmington) [inline pkgs/mkService.nix]: `{()}`

File: pkgs/mkService.nix [hunk 10]
```diff
@@ -113,7 +114,9 @@ let
     if user.name == "root"
       then exec
       else
-        let suf = lib.optionalString (user.groups != []) ":${lib.concatStringsSep ":" user.groups}";
+        let
+          allGroups = user.groups ++ user.userGroups;
+          suf = lib.concatMapStrings (g: ":${g}:") allGroups;
```
Comment (rimmington) [inline pkgs/mkService.nix]: No trailing `:`.

**Review by rimmington — CHANGES_REQUESTED** [review]

**Comment by rimmington** [comment]
@dwillie Changes look good. Do you want to squash those commits?

**Pull request merged**